lambda_calc/wrapper.py

```diff
@@ -0,0 +1,131 @@
+from typing import Generator, TypedDict, TypeAlias
+from functools import reduce
+import itertools
+import string
+
+from .core import alpha_equiv, get_env, all_beta_reductions, is_valid_reduction, is_simple
+from .parser import parse
+from .ast import Var, Fun, App
+
+
+class Reducer:
+    """
+    Wrapper class that holds all the information of the reduced lambda expression
+    
+    lambdastring: Lambda string to evaluate
+    candidatestring: prairelearn input string from candidate/student
+    """
+
+    alpha_error_list = []
+    beta_error_list = []
+    general_error_list = []
+
+    def __init__(self, lambdastring : str, candidatestring : str):
+        self.lambdastring = lambdastring
+        self.candidatestring = candidatestring
+        self.expr = parse(self.lambdastring)
+        self.all_beta_reducs = self.store_all_reducs()
+        self.candidate_beta_reducs = self.parse_candidate_str()
+        #Check the beta reductions
+        self.check_all_beta_reducs()
+
+    def store_all_reducs(self):
+        """
+        To store all the reductions
+        """
+        reducs_list = []
+        to_go = True
+        current_expr = self.expr
+        while to_go:
+            try:
+                curr_reduced = next(all_beta_reductions(current_expr))
+                reducs_list.append(curr_reduced)
+                current_expr = curr_reduced
+            except:
+                to_go = False
```

Comment:
this can probably be simplified to a list comprehension. and it looks like it only considers a single possible reduction path? what if the user inputs a different and valid path?

Reply:
I decided not to use list comprehension because we might not know if the current expression is in its simplest form.

On point 2,this function here just evaluates one path of the "correct" solution, but I would think that even if another path is considered, we can still check alpha equivalence? Alternatively, we could also iterate through all_beta_reductions them as a list in all_beta_reducs and change check_all_beta_reducs.